shed/tests/test_streams.py

```diff
@@ -546,6 +549,101 @@ def f(img1):
         assert n in assert_docs
 
 
+def test_filter_descriptor(exp_db, start_uid1):
+    source = Stream()
+
+    def f(d):
+        return d[0]['name'] == 'primary'
+
+    dp = es.filter(f, source,
+                   stream_name='test',
+                   document_name='descriptor')
+    L = dp.sink_to_list()
+    dp.sink(star(SinkAssertion(False)))
+
+    ih1 = exp_db[start_uid1]
+    s = exp_db.restream(ih1, fill=True)
+    for a in s:
+        source.emit(a)
+
+    assert_docs = set()
+    for l, s in zip(L, exp_db.restream(ih1, fill=True)):
+        assert_docs.add(l[0])
+        if l[0] == 'start':
+            assert l[1] != s[1]
+        if l[0] == 'event':
+            assert_allclose(l[1]['data']['pe1_image'],
+                            s[1]['data']['pe1_image'])
+        if l[0] == 'stop':
+            assert l[1]['exit_status'] == 'success'
+    for n in ['start', 'descriptor', 'event', 'stop']:
+        assert n in assert_docs
+
+    L.clear()
+    ih1 = exp_db[start_uid1]
+    s = exp_db.restream(ih1, fill=True)
+    for a in s:
+        source.emit(a)
+
+    assert_docs = set()
+    for l, s in zip(L, exp_db.restream(ih1, fill=True)):
+        assert_docs.add(l[0])
+        if l[0] == 'start':
+            assert l[1] != s[1]
+        if l[0] == 'event':
+            assert_allclose(l[1]['data']['pe1_image'],
+                            s[1]['data']['pe1_image'])
+        if l[0] == 'stop':
+            assert l[1]['exit_status'] == 'success'
+    for n in ['start', 'descriptor', 'event', 'stop']:
+        assert n in assert_docs
+
+
+def test_filter_descriptor_negative(exp_db, start_uid1):
+    source = Stream()
+
+    def f(d):
+        tv = d[0]['name'] != 'primary'
+        return tv
+
+    dp = es.filter(f, source,
+                   stream_name='test',
+                   document_name='descriptor')
+    L = dp.sink_to_list()
+    dp.sink(star(SinkAssertion(False, expected_docs={'start', 'stop'})))
+
+    ih1 = exp_db[start_uid1]
+    s = exp_db.restream(ih1, fill=True)
+    for a in s:
+        source.emit(a)
+
+    assert_docs = set()
+    for l, s in zip(L, exp_db.restream(ih1, fill=True)):
+        assert_docs.add(l[0])
+        if l[0] == 'start':
+            assert l[1] != s[1]
+        if l[0] == 'stop':
+            assert l[1]['exit_status'] == 'success'
+    for n in ['start', 'stop']:
+        assert n in assert_docs
+
+    L.clear()
+    ih1 = exp_db[start_uid1]
+    s = exp_db.restream(ih1, fill=True)
+    for a in s:
+        source.emit(a)
+
+    assert_docs = set()
+    for l, s in zip(L, exp_db.restream(ih1, fill=True)):
+        assert_docs.add(l[0])
+        if l[0] == 'start':
+            assert l[1] != s[1]
+        if l[0] == 'stop':
+            assert l[1]['exit_status'] == 'success'
+    for n in ['start', 'stop']:
+        assert n in assert_docs
```

Comment:
what about asserting that `'descriptor' not in assert_docs`?

Reply:
I really just do a set comparison, which I will fix.